lib/src/Uncertainty/Algorithm/Simulation/LineSampling.cxx

```diff
@@ -0,0 +1,284 @@
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/LineSampling.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/DistFunc.hxx"
+#include "openturns/ComposedFunction.hxx"
+#include "openturns/SpecFunc.hxx"
+#include "openturns/LinearFunction.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(LineSampling)
+
+static const Factory<LineSampling> Factory_LineSampling;
+
+
+/* Default constructor */
+LineSampling::LineSampling()
+  : EventSimulation()
+{
+}
+
+
+/* Constructor with parameters */
+LineSampling::LineSampling(const RandomVector & event,
+                           const Point & initialAlpha,
+                           const RootStrategy & rootStrategy)
+  : EventSimulation(event.getImplementation()->asComposedEvent())
+  , initialAlpha_(initialAlpha)
+  , rootStrategy_(rootStrategy)
+  , searchOppositeDirection_(ResourceMap::GetAsBool("LineSampling-DefaultSearchOppositeDirection"))
+  , adaptiveImportantDirection_(ResourceMap::GetAsBool("LineSampling-DefaultAdaptiveImportantDirection"))
+{
+  if (!event.isEvent() || !event.isComposite()) throw InvalidArgumentException(HERE) << "LineSampling requires a composite event";
+  UnsignedInteger outputDimension = getEvent().getFunction().getOutputDimension();
+  if (outputDimension > 1)
+    throw InvalidArgumentException(HERE) << "Output dimension for LineSampling cannot be greater than 1, here output dimension=" << outputDimension;
+  setInitialAlpha(initialAlpha);
+  standardEvent_ = StandardEvent(getEvent());
+}
+
+
+/* Virtual constructor */
+LineSampling * LineSampling::clone() const
+{
+  return new LineSampling(*this);
+}
+
+
+/* Performs the actual computation. */
+void LineSampling::run()
+{
+  // First, initialize some parameters
+  const UnsignedInteger dimension = initialAlpha_.getDimension();
+  alpha_ = initialAlpha_;
+  alpha_ /= alpha_.norm();
+  uStar_ = Point(dimension, std::sqrt(SpecFunc::MaxScalar));
+
+  // clear history
+  alphaHistory_ = Sample(0, dimension);
+  rootPointsHistory_.clear();
+  rootValuesHistory_.clear();
+
+  EventSimulation::run();
+}
+
+
+/* Compute the block sample */
+Sample LineSampling::computeBlockSample()
+{
+  const UnsignedInteger dimension = initialAlpha_.getDimension();
+  const UnsignedInteger blockSize = getBlockSize();
+  const Distribution inputDistribution(standardEvent_.getAntecedent().getDistribution());
```

Comment:
It will fail if the event is based on an antecedent which is also a CompositeVector. You cannot assume that the antecedent has a well-defined distribution (it is a long-standing bug). One way to overcome this is to check if the antecedent is also a CompositeVector in the constructor of the CompositeVector and to compose the functions. This way, the antecedent is always a RandomVector based on a distribution.

Reply:
done